src/main/java/com/google/codeu/data/Datastore.java

```diff
@@ -62,34 +63,95 @@ public void storeMessage(Message message) {
     datastore.put(messageEntity);
   }
 
-  /** Stores the user in Datastore (if not already in it) */
-  public void storeUser(String user) {
-    Entity userEntity = new Entity("User", user);
-    if (!userFound(user)) {
+  /**
+   * Either stores the user in datastore, or updates the messagesSent
+   * property when a message is being stored.
+   */
+  public void storeUser(String username, long messagesSent) {
+    Entity userEntity = new Entity("User", username);
+    userEntity.setProperty("messagesSent", messagesSent);
+    // Storing a user when they first log in.
+    if (messagesSent == 0) {
+      if (!userFound(username)) {
+        datastore.put(userEntity);
+      }
+    // Updating a user's messagesSent property here.
+    } else {
       datastore.put(userEntity);
     }
   }
 
+  /**
+   * Gets average message length.
+   * 
+   * @return average length of messages sent.
+   */
+  public int getAverageMessageLength() {
+    Query query = new Query("Message");
+    PreparedQuery results = datastore.prepare(query);
+    int numMessages = results.countEntities(FetchOptions.Builder.withLimit(fetchLimit));
+    // Use double for calculations and cast to int when returning.
+    double averageLength = 0;
+    for (Entity messageEntity: results.asIterable()) {
+      String text = (String) messageEntity.getProperty("text");
+      averageLength += (double) text.length() / numMessages;
```

Comment:
check divide by 0. Probably shouldn't happen because in theory the messages property is in sync with the message list, but there are scenarios where it could be off (eg, where one update succeeded and the other failed).